sys/net/application_layer/nanocoap/sock.c

```diff
@@ -81,53 +90,63 @@ ssize_t nanocoap_sock_request_cb(nanocoap_sock_t *sock, coap_pkt_t *pkt,
                                  coap_request_cb_t cb, void *arg)
 {
     ssize_t tmp, res = 0;
-    size_t pdu_len = (pkt->payload - (uint8_t *)pkt->hdr) + pkt->payload_len;
-    uint8_t *buf = (uint8_t*)pkt->hdr;
-    uint32_t id = coap_get_id(pkt);
+    const void *pdu = pkt->hdr;
+    const size_t pdu_len = coap_get_total_len(pkt);
+    const unsigned id = coap_get_id(pkt);
     void *payload, *ctx = NULL;
 
     unsigned state = STATE_SEND_REQUEST;
 
-    /* TODO: timeout random between between ACK_TIMEOUT and (ACK_TIMEOUT * ACK_RANDOM_FACTOR) */
-    uint32_t timeout = CONFIG_COAP_ACK_TIMEOUT_MS * US_PER_MS;
-
+    uint32_t timeout = random_uint32_range(CONFIG_COAP_ACK_TIMEOUT_MS * US_PER_MS,
+                                           CONFIG_COAP_ACK_TIMEOUT_MS * CONFIG_COAP_RANDOM_FACTOR_1000);
     /* add 1 for initial transmit */
     unsigned tries_left = CONFIG_COAP_MAX_RETRANSMIT + 1;
 
     /* check if we expect a reply */
     const bool confirmable = coap_get_type(pkt) == COAP_TYPE_CON;
 
+    /* try receiving another packet without re-request */
+    bool retry_rx = false;
```

Comment:
So this want's to handle the case when the wrong message ID is received, but the right one might be on its way or already in the queue, within the timeout window. But this could currently lead to doubling the timeout no? I think this is why in the old cold we had all this `deadline_left` code, I think something like that needs to be re-introduced for this.

Reply:
f67a819